File: src/main/java/com/yazan98/actor/processor/ActorProcessor.java
package com.yazan98.actor.processor;

import com.yazan98.actor.ActorConsts;
import com.yazan98.actor.annotations.Actor;
import com.yazan98.actor.annotations.ActorController;
import com.yazan98.actor.models.ActorApplicationInformation;
import com.yazan98.actor.models.ActorControllerInformation;

import javax.annotation.processing.AbstractProcessor;
import javax.annotation.processing.Messager;
import javax.annotation.processing.ProcessingEnvironment;
import javax.annotation.processing.RoundEnvironment;
import javax.lang.model.SourceVersion;
import javax.lang.model.element.Element;
import javax.lang.model.element.ElementKind;
import javax.lang.model.element.TypeElement;
import javax.lang.model.util.ElementFilter;
import javax.tools.Diagnostic;
import java.util.*;

public class ActorProcessor extends AbstractProcessor {

    private String defaultHost;
    private String defaultPort;
    private ArrayList<ActorControllerInformation> controllerInformations;
    private ActorApplicationInformation applicationInfo;
    private Set<String> supportedAnnotationTypes;
    private Messager messager;

    public ActorProcessor() {
        this.supportedAnnotationTypes = new HashSet<>();
        this.applicationInfo = new ActorApplicationInformation();
        this.supportedAnnotationTypes.add("com.yazan98.actor.annotations.Actor");
        this.controllerInformations = new ArrayList<>();
    }

    @Override
    public synchronized void init(ProcessingEnvironment processingEnv) {
        super.init(processingEnv);
        this.messager = processingEnv.getMessager();
    }

    @Override
    public boolean process(Set<? extends TypeElement> set, RoundEnvironment roundEnvironment) {
        Collection<? extends Element> annotatedElements = roundEnvironment.getElementsAnnotatedWith(Actor.class);
        List<TypeElement> types = ElementFilter.typesIn(annotatedElements);
        for (Element element : annotatedElements) {
            if (element.getKind() != ElementKind.CLASS) {
                messager.printMessage(Diagnostic.Kind.ERROR, ActorConsts.ACTOR_NOT_ON_CLASS);
                return false;
            } else {
                startExecuteAfterFirstValidate(types, roundEnvironment);
            }
        }
        return true;
    }

    private void startExecuteAfterFirstValidate(List<TypeElement> types, RoundEnvironment environment) {
        for (TypeElement type : types) {
            Actor applicationAnnotation = type.getAnnotation(Actor.class);
            if (applicationAnnotation.enabled()) {
                this.applicationInfo.setApplicationName(applicationAnnotation.application().getSimpleName() + ActorConsts.DEFAULT_NAME);
                this.applicationInfo.setDescription(applicationAnnotation.description());
                this.applicationInfo.setSchema(ActorConsts.DEFAULT_SCHEMA);
                this.applicationInfo.setId(UUID.randomUUID().toString());
                this.defaultHost = applicationAnnotation.serverHost();
                this.defaultPort = applicationAnnotation.serverPort();
                startExecuteControllerValidation(environment);
            } else {
                break;
            }
        }
    }

    private void startExecuteControllerValidation(RoundEnvironment environment) {
        Collection<? extends Element> annotatedElements = environment.getElementsAnnotatedWith(ActorController.class);
        List<TypeElement> types = ElementFilter.typesIn(annotatedElements);
        for (TypeElement element : types) {
            controllerInformations.add(new ActorControllerInformation());
        }
    }

    @Override
    public Set<String> getSupportedAnnotationTypes() {
        return supportedAnnotationTypes;
    }

    @Override
    public SourceVersion getSupportedSourceVersion() {
        return SourceVersion.RELEASE_8;
    }

}

File: src/main/java/com/yazan98/actor/processor/ActorProcessor.kt
package com.yazan98.actor.processor

import com.google.auto.service.AutoService
import com.yazan98.actor.ActorConsts
import com.yazan98.actor.annotations.ActorApplication
import com.yazan98.actor.annotations.ActorController
import com.yazan98.actor.annotations.ActorGetRequest
import com.yazan98.actor.models.ActorApplicationInfo
import com.yazan98.actor.models.ActorRequestInfo
import org.springframework.context.annotation.Configuration
import java.util.*
import javax.annotation.processing.*
import javax.lang.model.SourceVersion
import javax.lang.model.element.Element
import javax.lang.model.element.ElementKind
import javax.lang.model.element.TypeElement
import javax.lang.model.util.ElementFilter
import javax.tools.Diagnostic
import kotlin.collections.ArrayList

@SupportedAnnotationTypes(value = [
    "com.yazan98.actor.annotations.ActorApplication",
    "com.yazan98.actor.annotations.ActorController",
    "com.yazan98.actor.annotations.ActorGetRequest"
])
@SupportedSourceVersion(SourceVersion.RELEASE_8)
@AutoService(ActorProcessor::class)
open class ActorProcessor : AbstractProcessor() {

    private lateinit var messager: Messager
    private val applicationDetails: ActorApplicationInfo by lazy {
        ActorApplicationInfo()
    }

    private val controllers: ArrayList<ActorRequestInfo> by lazy {
        ArrayList<ActorRequestInfo>()
    }

    init {
        println("Processor Called")
    }

    override fun init(processingEnv: ProcessingEnvironment?) {
        super.init(processingEnv)
        processingEnv?.let {
            messager = it.messager
        }
    }

    override fun getSupportedAnnotationTypes(): MutableSet<String> {
        return mutableSetOf(
                ActorApplication::class.java.name,
                ActorController::class.java.name,
                ActorGetRequest::class.java.name
        )
    }

    override fun process(p0: MutableSet<out TypeElement>?, environment: RoundEnvironment?): Boolean {
        environment?.let {
            executeProjectAnnotation(it)
        }
        return true
    }

    private fun executeProjectAnnotation(environment: RoundEnvironment) {
        val annotatedElements: Collection<Element> = environment.getElementsAnnotatedWith(ActorApplication::class.java)
        val types = ElementFilter.typesIn(annotatedElements)
        for (element in annotatedElements) {
            if (element.kind != ElementKind.CLASS) {
                messager.printMessage(Diagnostic.Kind.ERROR, ActorConsts.ACTOR_NOT_ON_CLASS)
            } else {
                startExecuteAfterFirstValidate(types, environment)
            }
        }
    }

    private fun startExecuteAfterFirstValidate(types: List<TypeElement>, environment: RoundEnvironment) {
        for (type in types) {
            val applicationAnnotation: ActorApplication = type.getAnnotation(ActorApplication::class.java)
            if (applicationAnnotation.enabled) {
                this.applicationDetails.name = "${applicationAnnotation.name}-${ActorConsts.DEFAULT_NAME}"
                this.applicationDetails.description = applicationAnnotation.description
                this.applicationDetails.schema = (ActorConsts.DEFAULT_SCHEMA)
                this.applicationDetails.id = (UUID.randomUUID().toString())
                this.applicationDetails.host = applicationAnnotation.host
                this.applicationDetails.port = applicationAnnotation.port.toString()
                startExecuteControllerValidation(environment)
            } else {
                break
            }
        }
    }

    private fun startExecuteControllerValidation(environment: RoundEnvironment) {
        val annotatedElements: Collection<Element?> = environment.getElementsAnnotatedWith(ActorController::class.java)
        val types = ElementFilter.typesIn(annotatedElements)
        annotatedElements.forEach {
            if(it?.kind != ElementKind.CLASS) {
                messager.printMessage(Diagnostic.Kind.ERROR, ActorConsts.CONTROLLER_CLASS_ERROR)
            }
        }
        for (element in types) {
            val controller: ActorController = element.getAnnotation(ActorController::class.java)
            controllers.add(ActorRequestInfo(
                    controller.name,
                    controller.description
            ))
        }
        controllers.forEach {
            println("The Controller : $it")
        }
    }

    override fun getSupportedSourceVersion(): SourceVersion {
        return SourceVersion.RELEASE_10
    }

}
